<font size=6><b>Importing required packages:<b><font>

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
from sklearn.model_selection import train_test_split
import warnings
warnings.filterwarnings('ignore')


<font size=5><b>Reading data from dataset:<b><font>

In [2]:
df=pd.read_csv("balance-scale.csv")

In [3]:
df.head()


Class  L-Weight  L-Distance  R-Weight  R-Distance
0     B         1           1         1           1
1     R         1           1         1           2
2     R         1           1         1           3
3     R         1           1         1           4
4     R         1           1         1           5

In [4]:
df.describe()

L-Weight  L-Distance    R-Weight  R-Distance
count  625.000000  625.000000  625.000000  625.000000
mean     3.000000    3.000000    3.000000    3.000000
std      1.415346    1.415346    1.415346    1.415346
min      1.000000    1.000000    1.000000    1.000000
25%      2.000000    2.000000    2.000000    2.000000
50%      3.000000    3.000000    3.000000    3.000000
75%      4.000000    4.000000    4.000000    4.000000
max      5.000000    5.000000    5.000000    5.000000

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 625 entries, 0 to 624
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Class       625 non-null    object
 1   L-Weight    625 non-null    int64 
 2   L-Distance  625 non-null    int64 
 3   R-Weight    625 non-null    int64 
 4   R-Distance  625 non-null    int64 
dtypes: int64(4), object(1)
memory usage: 22.0+ KB


In [6]:
df.shape

(625, 5)

In [7]:
df.isnull().sum()

Class         0
L-Weight      0
L-Distance    0
R-Weight      0
R-Distance    0
dtype: int64

In [8]:
cols = ['L-Weight','L-Distance','R-Weight','R-Distance']
X = df[cols] 
y = df.Class 

In [9]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1) 
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(437, 4) (188, 4) (437,) (188,)


<font size=6><b>DecisionTreeClassifier Algorithm:<b><font>


In [10]:
from sklearn.tree import DecisionTreeClassifier 

dtc=DecisionTreeClassifier()


In [11]:
 dtc.fit(X_train,y_train)

DecisionTreeClassifier()

In [12]:
y_prediction = dtc.predict(X_test)

In [13]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, y_prediction)

array([[ 0, 10,  4],
       [ 9, 77,  4],
       [ 9,  7, 68]], dtype=int64)

In [14]:
from sklearn import metrics
print("Accuracy:",metrics.accuracy_score(y_test, y_prediction))

Accuracy: 0.7712765957446809


<font size=6><b>HyperParameter Tuning:<b><font>
<br><font size=3><b>using GridSearchCV<b><font>

In [15]:
param_i={ "criterion":["gini","entropy"],
         "max_depth":[1,2,3,4,5,6,7,None]
}

In [16]:
from sklearn.model_selection import GridSearchCV
grid=GridSearchCV(dtc,param_grid=param_i,cv=10,n_jobs=-1)

In [17]:
grid.fit(X_train,y_train)

GridSearchCV(cv=10, estimator=DecisionTreeClassifier(), n_jobs=-1,
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [1, 2, 3, 4, 5, 6, 7, None]})

In [18]:
grid.best_estimator_

DecisionTreeClassifier(criterion='entropy')

In [19]:
grid.best_score_

0.7758456659619449

In [20]:
grid.best_params_

{'criterion': 'entropy', 'max_depth': None}

<h2><strong>RandomForestRegressor</strong></h2>

In [42]:
from sklearn.ensemble import RandomForestRegressor


<h5><strong>One Hot Encoder:</strong></h5>

In [43]:
from sklearn.preprocessing import OneHotEncoder
enc = OneHotEncoder(handle_unknown='ignore')
enc_df=pd.DataFrame(enc.fit_transform(df[['Class']]).toarray())
enc_df


0    1    2
0    1.0  0.0  0.0
1    0.0  0.0  1.0
2    0.0  0.0  1.0
3    0.0  0.0  1.0
4    0.0  0.0  1.0
..   ...  ...  ...
620  0.0  1.0  0.0
621  0.0  1.0  0.0
622  0.0  1.0  0.0
623  0.0  1.0  0.0
624  1.0  0.0  0.0

[625 rows x 3 columns]

In [44]:
cols = ['L-Weight','L-Distance','R-Weight','R-Distance']
X = df[cols]

y = enc_df
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1) 
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(437, 4) (188, 4) (437, 3) (188, 3)


In [45]:
regr= RandomForestRegressor(n_estimators = 10, random_state = 0)
regr.fit(X_train,y_train)

RandomForestRegressor(n_estimators=10, random_state=0)

In [47]:
from sklearn.metrics import mean_squared_error
y_pred = regr.predict(X_test)


MSE= mean_squared_error(y_test, y_pred)
print(MSE)

0.08985815602836877


In [49]:
score = regr.score(X_train, y_train)
score

0.8881873494135601

In [50]:
regr.get_params()

{'bootstrap': True,
 'ccp_alpha': 0.0,
 'criterion': 'mse',
 'max_depth': None,
 'max_features': 'auto',
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_impurity_split': None,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 10,
 'n_jobs': None,
 'oob_score': False,
 'random_state': 0,
 'verbose': 0,
 'warm_start': False}

The MEAN SQUARED ERROR is 0.08985815602836877.
Thus, it can be concluded that the parameters for the model is sufficient enough for the model to perform prediction.